api/errors.go

```diff
@@ -10,39 +11,96 @@ import (
 // The canonical errors from the EigenDA gRPC API endpoints.
 //
 // Notes:
-// - We start with a small (but sufficient) subset of grpc's error codes,
-//   and expand when there is an important failure case to separate out. See:
-//   https://grpc.io/docs/guides/status-codes/
-// - Make sure that internally propagated errors are eventually wrapped in one of the
-//   user-facing errors defined here, since grpc otherwise returns an UNKNOWN error code,
-//   which is harder to debug and understand for users.
-// - See https://github.com/googleapis/googleapis/blob/ba8ea80f25d19bde8501cd51f314391f8d39bde8/google/rpc/code.proto
-//   for the mapping of grpc error codes to HTTP status codes.
-
+//   - We start with a small (but sufficient) subset of grpc's error codes,
+//     and expand when there is an important failure case to separate out. See:
+//     https://grpc.io/docs/guides/status-codes/
+//   - Make sure that internally propagated errors are eventually wrapped in one of the
+//     user-facing errors defined here, since grpc otherwise returns an UNKNOWN error code,
+//     which is harder to debug and understand for users.
+//   - See https://github.com/googleapis/googleapis/blob/ba8ea80f25d19bde8501cd51f314391f8d39bde8/google/rpc/code.proto
+//     for the mapping of grpc error codes to HTTP status codes.
 func newErrorGRPC(code codes.Code, msg string) error {
 	return status.Error(code, msg)
 }
 
+// WrapAsGRPCError wraps an error with a gRPC code and message. If the error already has a gRPC code,
+// it retains the code but combines the messages. Otherwise, it uses the provided code and combined message.
+func WrapAsGRPCError(err error, code codes.Code, msg string) error {
+	if err == nil {
+		return nil
+	}
+
+	combinedMessage := fmt.Sprintf("%s: %s", msg, err.Error())
+	if s, ok := status.FromError(err); ok {
+		// If it's already a gRPC error, use the same code but combine the messages
+		return status.Error(s.Code(), combinedMessage)
+	}
+
+	// If it's not a gRPC error, create a new gRPC error with the given code and message
+	return status.Error(code, combinedMessage)
+}
+
+// WrapGRPCError wraps an error with a message. If the error is a gRPC error, it retains the code
+// and combines the messages. Otherwise, it returns a new error with the combined message.
+func WrapGRPCError(err error, msg string) error {
+	if err == nil {
+		return nil
+	}
+
+	combinedMessage := fmt.Sprintf("%s: %s", msg, err.Error())
+	if s, ok := status.FromError(err); ok {
+		// If it's a gRPC error, keep it as gRPC error but combine the messages
+		return status.Error(s.Code(), combinedMessage)
+	}
+
+	return errors.New(combinedMessage)
```

Comment:
this function has weird semantics, its called WrapGRPCError but can accept a non-grpc error and return a non-grpc error. Can we simplify it instead?

Reply:
That's exactly the desired semantics. If there is a better concise name I'm happy to use it.

Reply:
WrapGRPCError indicates that only GRPC errors should be allowed to be wrapped, yet your documentation says that its fine to pass another error and itll just get wrapped normally. Error wrapping also typically means wrapping an error, not just extending the error with some other msg string. So it feels convoluted and unclear to me.

I like when functions have a single small, and clear purpose. Just like golang's errors api, only has .Is() and .As(), and errors.join() for combining errors, thats it. Forces users to use it in the same way, and not too much cognitive overload when trying to use it.

How about we mention that this should ONLY be used for wrapping grpc errors? I don't think any function should return either a GRPC error or not right?

Reply:
golang's api needs to be small as it has no context

the functions are not expected to return grpc error, only the APIs do; and the wrapper here works no matter if a function returns grpc error or not.